sound/soc/sof/intel/hda-dai.c

```diff
@@ -187,46 +187,41 @@ static int hda_dai_widget_update(struct snd_soc_dapm_widget *w,
 	return hda_ctrl_dai_widget_free(w, SOF_DAI_CONFIG_FLAGS_NONE, &data);
 }
 
-static int hda_dai_hw_params_update(struct snd_pcm_substream *substream,
-				    struct snd_pcm_hw_params *params,
-				    struct snd_soc_dai *dai)
-{
-	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(dai, substream->stream);
-	const struct snd_sof_dai_ops *ops = hda_dai_get_ops(substream, dai);
-	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(dai->component);
-	struct hdac_ext_stream *hext_stream = NULL;
-	int stream_tag;
-
-	if (ops && ops->dma_ops && ops->dma_ops->stream_get_dma)
-		hext_stream = ops->dma_ops->stream_get_dma(sdev, dai, substream);
-	if (!hext_stream)
-		return -EINVAL;
-
-	stream_tag = hdac_stream(hext_stream)->stream_tag;
-
-	/* set up the DAI widget and send the DAI_CONFIG with the new tag */
-	return hda_dai_widget_update(w, stream_tag - 1, true);
-}
-
 static int hda_dai_hw_params(struct snd_pcm_substream *substream,
 			     struct snd_pcm_hw_params *params,
 			     struct snd_soc_dai *dai)
 {
+	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(dai, substream->stream);
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(dai->component);
 	const struct snd_sof_dai_ops *ops = hda_dai_get_ops(substream, dai);
+	const struct sof_ipc_tplg_ops *tplg_ops = sdev->ipc->ops->tplg;
+	struct snd_sof_widget *swidget = w->dobj.private;
 	struct hdac_ext_stream *hext_stream = NULL;
+	struct snd_sof_dai_config_data data;
 	int ret;
 
 	if (ops && ops->dma_ops && ops->dma_ops->stream_get_dma)
 		hext_stream = ops->dma_ops->stream_get_dma(sdev, dai, substream);
+
 	if (hext_stream && hext_stream->link_prepared)
 		return 0;
 
 	ret = hda_link_dma_hw_params(substream, params, dai);
 	if (ret < 0)
 		return ret;
 
-	return hda_dai_hw_params_update(substream, params, dai);
+	if (ops && ops->dma_ops && ops->dma_ops->stream_get_dma)
+		hext_stream = ops->dma_ops->stream_get_dma(sdev, dai, substream);
+
+	if (tplg_ops && tplg_ops->dai_config) {
+		unsigned int flags = SOF_DAI_CONFIG_FLAGS_HW_PARAMS;
+
+		flags |= SOF_DAI_CONFIG_FLAGS_2_STEP_STOP << SOF_DAI_CONFIG_FLAGS_QUIRK_SHIFT;
```

Comment:
also here for data.dai_index init ?

Reply:
dai_index is only needed for SDW and it doesnt use these ops